### DISCUSSION TITLE
Oxalis upgrade

### DISCUSSION THREAD

**Discussion by swouters1** [OxalisCommunity/oxalis#704] [General]
Can someone help me or give some advice?

We’re trying to update our oxalis code to the latest version.  It works fine for inbound but we’re having some performance issues for outbound.
We can reproduce the issue with just running the standalone jar file on the command line.
See ‘Total time spent’ for both cases and you will see it takes much longer for us when using the latest version.

1)	When we send the test message to the Peppol test bed and use Java 11, Oxalis 6.7 JARs using the following command

java -classpath "oxalis-standalone.jar:$MS_JAVA/external/oxalis-as4-6.7.0-dist/*
" eu.sendregning.oxalis.Main \
  -f /home/bes/ms/as4.test.xml \
  -s 9932:xxxxxxxxxxxxxx \
  -r 9922:xxxxxxxxxxxxxx \
  -t \
  --protocol peppol-transport-as4-v2_0

Selected excerpts from the log
…
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.sun.xml.bind.v2.runtime.reflect.opt.Injector (file:/home/bes/ms/oxalis-standalone.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int)
WARNING: Please consider reporting this to the maintainers of com.sun.xml.bind.v2.runtime.reflect.opt.Injector
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
2024-11-15 10:05:42,894 INFO [network.oxalis.commons.filesystem.detector.EnvironmentHomeDetector] Using Oxalis folder specified as environment variable 'OXALIS_HOME' with value '/usr1/.oxalis'.
…
Nov 15, 2024 10:10:34 AM org.apache.cxf.configuration.jsse.SSLUtils loadDefaultKeyManagers
WARNING: Default key managers cannot be initialized: Password must not be null
java.security.UnrecoverableKeyException: Password must not be null
        at java.base/sun.security.provider.JavaKeyStore.engineGetKey(JavaKeyStore.java:135)
…
2024-11-15 10:10:39,957 WARN [org.apache.wss4j.common.crypto.CryptoBase] No Subject DN Certificate Constraints were defined. This could be a security issue
31ecdee5-a04b-454a-a6b8-8d09a9db7418@vm4Dbes27 transmission took 11222 ms
Average transmission time was 11222.0msAverage transmission time was 11222.0ms
Total time spent: 284s
Attempted to send 1 files
Failed transmissions: 0
Transmission speed 0 documents per second 

2)	When we send the test message to the Peppol test bed and use Java 8, Oxalis 5.1 JARs using the following command:

java -classpath "blah/oxalis-standalone-as4/vefa-peppol/*:blah/oxalis-standalone-as4/standalone/*:blah /oxalis-stand
alone-as4/as4/*:$CLASSPATH" eu.sendregning.oxalis.Main \
-f /home/bes/ms/as4.test.xml \
  -s 9932:xxxxxxxxxxxxxx \
  -r 9922:xxxxxxxxxxxxxx \
  --protocol peppol-transport-as4-v2_0

Selected excerpts from the log
…
2024-11-15 11:09:08,782 WARN [org.apache.wss4j.common.crypto.CryptoBase] No Subject DN Certificate Constraints were defined. This could be a security issue
87e746ad-abaf-4334-9042-d81fc7b17771@vm4Dbes27 transmission took 7558 ms
2024-11-15 11:09:03,529 WARN [org.apache.cxf.ws.policy.AssertionBuilderRegistryImpl] No assertion builder for type {http://oxalis.network/custom/security-policy}Basic128GCMSha256MgfSha256 registered.
2024-11-15 11:09:05,177 WARN [org.apache.cxf.transport.https.SSLUtils] Default key managers cannot be initialized: Password must not be null
java.security.UnrecoverableKeyException: Password must not be null
        at sun.security.provider.JavaKeyStore.engineGetKey(JavaKeyStore.java:132)
…
2024-11-15 11:09:08,782 WARN [org.apache.wss4j.common.crypto.CryptoBase] No Subject DN Certificate Constraints were defined. This could be a security issue
87e746ad-abaf-4334-9042-d81fc7b17771@vm4Dbes27 transmission took 7558 ms
Average transmission time was 7558.0msAverage transmission time was 7558.0ms
Total time spent: 39s
Attempted to send 1 files
Failed transmissions: 0
Transmission speed 0 documents per second

Do you have any idea why there is such a big difference?

**Comment by aaron-kumar** [comment]
Thank you for sharing your feedback. However, we believe that the statistics provided and the methodology used to measure the performance of mentioned two versions of Oxalis may not be entirely accurate. Benchmark testing should always be conducted in a controlled, isolated environment to avoid the influence of external factors.

It is important to note that overall delivery performance can also be affected by various components such as Access Point integration, the load balancer, database, and queuing systems.

Based on the data we have gathered for the mentioned versions, we found that the performance of the latest version of Oxalis is either on par with or exceeds the previous version. However, considering the anticipated increase in traffic, which will be further augmented by adoption of MLS in near future, we are planning to conduct a more comprehensive performance and benchmark testing soon (_Possibly Q1 2025 - to be agreed by OMT_).

For further discussion, please refer to the following internal roadmap issue: https://github.com/OxalisCommunity/Oxalis-internal-roadmap/issues/30 (access restricted to Oxalis Members)

**Reply by swouters1** [reply]
thank you for taking some time to respond.  
I'm not sure why you think the performance test is not accurate.
 
We repeated the same test many times.  We are using the same sender and receiver IDs and run on the same server.  The only difference is the upgrade of the oxalis software and upgrade of java version.
Every time the transfer goes quick for the old version and it takes a long time for the new version.  The timings from the log are quite consistent and the timings in the logs are typical of each.

**Comment by aaron-kumar** [comment]
Converting it to discussion in favor of https://github.com/OxalisCommunity/Oxalis-internal-roadmap/issues/30
